### ISSUE TITLE
Setting useRaf=false in config results in exception

### ISSUE THREAD

**Issue by lawrence-dol** [leeoniya/domvm#57]
Another problem which does not show in a simple fiddle.

My application does not use routes. When I add `domvm.view.config({ useRaf: false });` right before mounting the root view, I get an "impossible" exception on first redraw:

````
19:42:23.807 TypeError: vm.render is null
redraw() AppScript;01:105
receiveData() AppScript;01:595
init() AppScript;01:592
MdlHostReplay() AppScript;01:605
view() AppScript;01:544
createView() AppScript;01:92
init() AppScript;01:542
AppMain() AppScript;01:549
<anonymous> App:24
1 AppScript;01:105:86
````

Inspecting the vm object, it does indeed not contain a render function. All I need to do to fix things is remove that line. (I only added it because your fiddle had it and I was checking to see if it impacted #56.)

Also, it doesn't appear to matter what you supply for config; even an empty object causes the exception.

**Comment by leeoniya** [comment]
yes, very strange. the only way `vm.render` can stay null is if your view doesnt return a `render()` [1]. `domvm.view.config()` has no impact on this.

[1] https://github.com/leeoniya/domvm/blob/1.x-dev/src/view.js#L141

**Comment by leeoniya** [comment]
@lawrence-dol

any way to reproduce this?

**Comment by lawrence-dol** [comment]
I'll look at it further soon. Right now I am making up a little for lost 
time. You can assign it to me to investigate, if you'd like.

**Comment by lawrence-dol** [comment]
BTW, this is still occurring.

**Comment by leeoniya** [comment]
...and i suspect it will continue to occur at least until you give me a reproducible test case ;)

**Comment by lawrence-dol** [comment]
> and i suspect it will continue to occur at least until...

Right you are.

This is actually the result of my model initialization calling using a custom property get/set which is doing a redraw... which redraw is then being invoked during initialization of the view-model while render is not yet set (obviously, because the view function has not yet returned).

There are a few ways I can work around this, but I wonder if this will be a common-enough scenario that it's worth putting a guard in redraw, checking if render is set?

A little more interesting is why this only happens when setting config (not even specifically `useRaf`, but even an empty config).

It seems to be something of a catch-22. I need the v-m to create my data model, since many items in it need to trigger redrawing on change (whether the way I've done it, or via a watcher), but I can't then change any properties in my model until after v-m init is complete. 

Maybe, being in the deep of a big app, I'm overlooking something simple? I will see if I can knock up a simple example as a fiddle.

**Comment by lawrence-dol** [comment]
Not having any luck with a fiddle; I will need to be more precise in exactly where it's failing (it's definitely a result of my code invoke `redraw`, though).

**Comment by lawrence-dol** [comment]
Little more progress. Setting a blank config clearly causes `useRaf` to default to false and the problem is only tripped when `useRaf==false`. That surely means that the reason is an inline redraw vs. a redraw on the next animation frame, no?

**Comment by leeoniya** [comment]
> It seems to be something of a catch-22. I need the v-m to create my data model, since many items in it need to trigger redrawing on change...

I'm not sure how this is a problem. You *have* the vm already and can pass it around / dep-inject it as needed. Yes, it's uninitialized so you cannot call its redraw until everything initializes, but you should never need to do this. Why you're doing it in your case remains a mystery.

**Comment by leeoniya** [comment]
> That surely means that the reason is an inline redraw vs. a redraw on the next animation frame, no?

yes, that's likely the culprit.

**Comment by lawrence-dol** [comment]
> Why you're doing it in your case remains a mystery.

I am in the process of pinning it down precisely, but the simple answer is DRY code. I have host data from which a whole bunch of very complex model values need to be derived. The easiest way to get the model initialized is to call the code that derives the model from an initial default "input" data. That way it's guaranteed that the derived state be initially correct, because it's manipulated in the same way that subsequent data received from the host will be.

Make sense?

**Comment by leeoniya** [comment]
> Make sense?

Eh, not as much as some code would :/

What doesnt make sense is why the model's init is intrinsically tied to rendering it out. All this model data can presumably be computed and stored without any reliance on a view, no?

**Comment by lawrence-dol** [comment]
> All this model data can presumably be computed and stored without any reliance on a view, no?

That would be true were it not that changes to the model values must trigger redraws to update the views. There's absolutely no ties to how the model is rendered, just properties that have the effect that "if my value is changed we have to redraw".

I'm open to better suggestions.

**Comment by lawrence-dol** [comment]
This is, I think, a perfectly legitimate use case. I'll explain with some code with the non interesting bits abstracted away.

Ultimately, there will be three separate pieces of code:

  1. A module that interacts with the host over the network.
  2. A module that can replay recorded data received from the host.
  3. A module that will do all the stuff common to both.

The difference between 1 and 2 is solely how they obtain a data block and what they do when the use submits data.

````
// genU.prop is just a getter/setter that can take a function to invoke when the 
// value is changed; in this case the function is redraw. The data model is 
// defined and frozen to prevent accidental changes by app code. It must 
// represent the host data and only that. Not transient processing state.
model: Object.freeze({
    dspData=Object.freeze({
        idn             : genU.prop(undefined),
        text            : genU.prop([]),
        cellClasses     : genU.prop([]),
        dspRow          : genU.prop(24),
        dspCol          : genU.prop(80),
        csrRow          : csrrow,
        csrCol          : csrcol,
        fldList         : genU.prop([]),
        augList         : genU.prop([]),
        enhList         : genU.prop([]),
        properties      : genU.prop({}),
        });

    stsData=Object.freeze({
        attribute       : genU.prop(""   ,appVm.redraw),
        csrRow          : csrrow                       ,
        csrCol          : csrcol                       ,
        message         : genU.prop(""   ,appVm.redraw),
        messageOverride : genU.prop(""   ,appVm.redraw),
        sessionName     : genU.prop(""   ,appVm.redraw),
        sessionIndex    : genU.prop(0    ,appVm.redraw),
        connectedFlag   : genU.prop(false,appVm.redraw),
        fileFlag        : genU.prop(false,appVm.redraw),
        insertFlag      : genU.prop(false,appVm.redraw),
        inputFlag       : genU.prop(false,appVm.redraw),
        messageFlag     : genU.prop(false,appVm.redraw),
        signonFlag      : genU.prop(false,appVm.redraw),
        shareFlag       : genU.prop(false,appVm.redraw),
        textFlag        : genU.prop(false,appVm.redraw),
        signoffFlag     : genU.prop(false,appVm.redraw),
        progressFlag    : genU.prop(false,appVm.redraw),
        endedFlag       : genU.prop(false,appVm.redraw),
        });
    });

// The recording replay module must initialize the model like so:
    stsData.message("Replaying recorded screens.");
    stsData.connectedFlag(true);
    stsData.sessionName  ("Recording");
    stsData.sessionIndex (0);
    receiveData(/*initial dummy "host" data*/);

// The host communications module must initialize the model like so:
    stsData.message("Not connected.");
    stsData.connectedFlag(false);
    stsData.sessionName  ("");
    stsData.sessionIndex (0);
    receiveData(/*initial dummy "host" data*/);

// The receive data function manipulates the data to initialize the model
// The decoding functions perform many pre-optimizations and derivations on the
// data, so that the application can use it without repeatedly redoing those.
function receiveData(hstdta) {
    hstData=hstdta;

    var dsptpl          =decodeDisplayData(hstdta.DisplayData);
    var dsprow          =dsptpl[0].length;
    var dspcol          =dsptpl[0][0].length;
    var dspfld          =decodeFields(hstdta.Field,dsptpl[0],dsptpl[1],dspcol);

    dspData.idn         (hstDataIndex);
    dspData.text        (dsptpl[0]);
    dspData.cellClasses (dsptpl[1]);
    dspData.dspRow      (dsprow);
    dspData.dspCol      (dspcol);
    dspData.csrRow      (hstdta.CursorRow          || 1);
    dspData.csrCol      (hstdta.CursorColumn       || 1);
    dspData.fldList     (dspfld);
    dspData.augList     (hstdta.AugObjects         || []);
    dspData.enhList     (hstdta.EnhObjects         || []);
    dspData.properties  (hstdta.SessionProperties  || {});

    stsData.sessionIndex (hstDataIndex+1);
    stsData.inputFlag    (hstdta.IndicatorFlags[0]=="1");
    stsData.messageFlag  (hstdta.IndicatorFlags[1]=="1");
    stsData.signonFlag   (hstdta.IndicatorFlags[2]=="1");
    stsData.shareFlag    (hstdta.IndicatorFlags[3]=="1");
    stsData.textFlag     (hstdta.IndicatorFlags[4]=="1");
    stsData.signoffFlag  (hstdta.IndicatorFlags[5]=="1");
    stsData.progressFlag (hstdta.IndicatorFlags[6]=="1");
    stsData.endedFlag    (hstdta.IndicatorFlags[7]=="1");

    dvm.u.tick(function() { appVm.redraw(); },1);
    }
````

Both setting the initial status values and of processing the initial prefab host data result in redraw being invoked. Yet it's the proper separation of concerns, IMO.

**Comment by leeoniya** [comment]
There's a lot of extra stuff to look through here, for example you could do away with 85% of the duplicate noise. I understand that it's easy to copy/paste your actual code, but really i need a reduced runnable scenario without having to imagine the remaining operational intricacies. You may have a valid use case here, but unfortunately it's not distilled enough and additional explaining will not be sufficient. We both understand you're redrawing before the vm is initialized, which may indicate that this chicken-egg issue may not be one that's solvable by the lib, i dunno.

I am certain that it is possible for you to write a clean 20 line demo that exactly mimics your architecture/init pattern in this case. I cannot help you until you do this.

**Comment by lawrence-dol** [comment]
Fair enough; I know it is frustrating to have someone just code dump on top of you. I was simply demonstrating that I have a real data model with significant complexity that makes it not a case of "just do the init differently from the normal data processing".

I will put together a fiddle as soon as I can to demonstrate the problem. It may be that in the process I'll realize where things went awry. I may need to just decouple the creation of the model until after the view-model is created.

But it would be really nice if the requests to redraw could somehow be decoupled from actually needing access to the `vm` object in question or if the DOMVM could simple ignore redraws on a vm until the constructor has returned the render function. It's a real pain-point to have special circumstances to bear in mind during construction *when it's natural to do a lot of work at that point*.

**Comment by leeoniya** [comment]
>  ... or if the DOMVM could simple ignore redraws on a vm until the constructor has returned the render function.

you're the one invoking redraw on the vm you have access to. what stops you from just adding a simple `if (vm.render)` check as a condition for redraw?

is this sufficient to solve your case?

adding it natively to domvm may present a problem when the vm in question is a subview because the vm gets replaced by the vtree branch that render() returns.

**Comment by lawrence-dol** [comment]
> what stops you from just adding a simple if (vm.render) check as a condition for redraw?

Nothing. But that simply ends up baked into a standard wrapping function, because it's not feasible to constantly ask yourself whether this is one of "those" properties that is set before the vm is ready. It seems like the kind of check that redraw could do internally and doing so would remove an edge-case (and not a specifically uncommon one at that).

**Comment by leeoniya** [comment]
i'm not convinced this is a common requirement; i've written a lot of code and never once needed it. if you provide a convincing demo, i will certainly consider it.

it's not as trivial to add as it seems. internal subview vtree branches get created by recursive redraw/render at the time of template -> vtree conversion. if a subview's redraw does not result in a valid branch, many things would go wrong. at the root level this is okay, but not any deeper. having different behavior for views is not something i want.

**Comment by lawrence-dol** [comment]
Okay, I get that. Trying to refactor the dependencies is not working either. If a model needs to trigger a redraw on a view in which it's used it that might happen during initialization redraw has to be guarded during initialization.

I have other ideas, but I'll need to sleep on it... my brain's done for the day. I am beginning to suspect that my model's doing too much.

**Comment by lawrence-dol** [comment]
Here's a question, though. Is there any real reason to have `useRaf` false? That is, I never expect redraw to be inline and immediate (in fact, I wrapped Mithril's to force it to occur on the next event loop cycle). I am perfectly happy to avoid the problem altogether by always using request animation frame.

I am not hung up on frame-rates; I expect a redraw request to be exactly that, just a request, and for multiple requests in the same event cycle to be automatically coalesced. I certainly don't want redraw work being done every time I change one of 30 properties as a result of an incoming data frame.

**Comment by leeoniya** [comment]
the primary reason for useRaf/false is accurate timing for perf and detect regressions and ensure the lib internally isnt for some reason triggering redraw more than it needs to. basically it's a debugging tool.

however, i would not recommend relying on tick/rAF behavior to ensure your vms are initialized...though this may be guaranteed by javascript's nature.

you should be able to trigger redraw from within models, it's an obvious use-case. but triggering it from model init (which itself is a dependency-injection for the view) makes no logical sense to me.

do what works for you, though. i'm still interested in a demo of what you're doing even if you just end up relying on rAF/tick.

**Comment by lawrence-dol** [comment]
Here's a crude demo: https://jsfiddle.net/LawrenceDol/yjyhm9se/

Switch between safeRedraw and unsafeRedraw in Model; unsafe will fail. 

Interestingly `w.prop` works, but I can't see in the code where it fires the handlers on property set, so that might be why. However I was under the impression that any change to a w.prop would cause a redraw on the related view.

**Comment by leeoniya** [comment]
because initializing a `w.prop` does not fire the handler, only mutating it does.

```js
var x = w.prop("foo");   // does nothing
x("bar");  // fires handler
```

**Comment by lawrence-dol** [comment]
> triggering it from model init (which itself is a dependency-injection for the view) makes no logical sense to me.

But you can't inject a model into the view if the model needs to use `vm.redraw`?? Regardless of whether you trigger redraw during it's init. If a model needs to trigger redraw and it uses the model, it must, by definition be constructed within the view init.

**Comment by lawrence-dol** [comment]
> because initializing a w.prop does not fire the handler, only mutating it does.

Yes, but my fiddle does set a `w.prop` during init.

**Comment by leeoniya** [comment]
you can invoke redraw from anywhere after the view closure has returned.

lemme look more. it's late.

**Comment by lawrence-dol** [comment]
I think there's a disconnect in my expectations. I wasn't expecting `redraw` to actually invoke the internal redraw logic inline, only to queue up a request to redraw on the next event-loop cycle. That all drawing internals would hang off an event, entirely separate from my code. That might well be an entirely unreasonable expectation. As result of too much time spent with GUI toolkits where "repaint" is always a request to paint, not an actual paint invocation.

**Comment by leeoniya** [comment]
probably the reason that w.prop works is cause it uses `w.fire` (which is internally rAF-debounced) to trigger the handler.

one thing that may help here and would be easy to do is if `w.prop` could be initialized empty. and the first mutation be considered an init and would avoid redraw. might be too magic, though.

**Comment by lawrence-dol** [comment]
Way too fragile, I think.

Why not just differentiate between "redraw" and "render"? Redraw queues a request and render invokes the render functions? IOW, redraw is always deferred to the next event loop? That would be more intuitive to me, at least.

**Comment by leeoniya** [comment]
ok. time for bed. more tomorrow.

if you invert the model/view relationship, btw: https://jsfiddle.net/yjyhm9se/13/

**Comment by leeoniya** [comment]
after some poking, there seem to be the same issues when the model/view is inverted, too.

it broke no tests and is pretty simple, so i added it tentatively to https://github.com/leeoniya/domvm/tree/no-redraw-uninit-vms branch: https://jsfiddle.net/yjyhm9se/15/

let me know if that works out.

**Comment by leeoniya** [comment]
> Why not just differentiate between "redraw" and "render"? Redraw queues a request and render invokes the render functions?

this is not gonna happen. the duality will introduce more confusion than problems it solves.

**Comment by lawrence-dol** [comment]
> Never going to happen.

Just a thought. I'll use my own as I did for Mithril. DOMVM makes that an easy path to implement.

**Comment by lawrence-dol** [comment]
So the test branch breaks after the second view reconstruction or a re-render. I am happy to simply use an app level "request redraw" wrapper if you don't want to spend time on this issue. Or I am happy to help if you do.

Just FYI, the exception is:

````
19:09:40.622 TypeError: vm.render.call is not a function
redraw() AppScript;01:106
createView/vm.update() AppScript;01:91
procProps() AppScript;01:224
procNode() AppScript;01:210
initNode() AppScript;01:137
initNode() AppScript;01:141
initNode() AppScript;01:141
initNode() AppScript;01:141
redraw() AppScript;01:107
moveTo() AppScript;01:97
buildNode/<() AppScript;01:161
forEach() self-hosted:208
buildNode() AppScript;01:159
redraw() AppScript;01:114
redraw() AppScript;01:104
getset() AppScript;01:496
submit() AppScript;01:600
prevScreen() AppScript;01:562
invokeHandler() AppScript;01:558
keyDownEvent() AppScript;01:557
wrapHandler/handler() AppScript;01:215
1 AppScript;01:106:100
````

**Comment by leeoniya** [comment]
i'd like to support this case natively if it doesnt break uniformity, cost a lot of bytes or require major restructuring.

at first, it appears to be something that can be handled without too much trouble.

**Comment by lawrence-dol** [comment]
OK. I've got to stop working for the weekend, but will be back on Monday. (Did you really work all night last night?)

**Comment by leeoniya** [comment]
hard to go to bed with a problem half-solved...

**Comment by lawrence-dol** [comment]
I hear that.

**Comment by leeoniya** [comment]
@lawrence-dol ping

**Comment by lawrence-dol** [comment]
Been off work for a bit and am working on another project at this moment. Last thing I did was try the `no-redraw-uninit-vms` which threw the exception reported above. But I will double check either today or tomorrow and report back.

**Comment by lawrence-dol** [comment]
@leeoniya : Using the alternate branch I do indeed get an exception still, but only after a number of redraws (4 separate event cycles, AFAICT).

The initial creation is followed by 19 "redraws", before the initial display, all of which are changes to properties in the model which result in inline redraw calls within the same event handler.

Then I type a key twice in an input, and on the third one it throws an exception as follows. 

I will see if I can reproduce this in a simple fiddle.


````
15:56:15.924 [**] recreated view AppScript:156:299
15:56:15.925 [**] render view (19) AppScript:156:352
15:56:25.406 [**] render view (3) AppScript:156:352
15:56:46.587 TypeError: p.render.call is not a function AppScript:17:100

Stack trace:

TypeError: p.render.call is not a function
m() AppScript:17
m/p.update() AppScript:18
q() AppScript:24
q() AppScript:25
q() AppScript:25
q() AppScript:25
m() AppScript:17
u() AppScript:16
x/<() AppScript:27
forEach() self-hosted:232
x() AppScript:26
m() AppScript:17
getset() AppScript:80
keyDownEvent() AppScript:141
A/<() AppScript:30
````

DOMVM begins at line 12 of my app script concatenation, so all line numbers need -11.

````
// Copyright (c) 2013 Advanced BusinessLink PTY/LTD, Australia.
// Copyright (c) 2013 Advanced BusinessLink Corporation, USA.


/* File 1: AppLeader.js */
var TemJS=(function(){

/* File 2: domvm.min.js */
/*
 https://github.com/leeoniya/domvm (1.x-dev)
*/
("undefined"!==typeof global? ...
````

**Comment by lawrence-dol** [comment]
Here is a fiddle based on that branch that doesn't crash with the exception but does loose input on the two properties that don't use `domvm.watch` because of the counter-intuitive action of updating the views inline.

    https://jsfiddle.net/LawrenceDol/yjyhm9se/

I will branch this and work further to reproduce the actual exception I see and figure out why the behavior in jsfiddle is different from what I see in my app running directly in the browser.

**Comment by lawrence-dol** [comment]
I am closing this, because it seems to be a problem that is inherent to inline redraw, which I am eschewing entirely for much broader reasons, and it's not worth my time to work out why my, much more complex, app is encountering this when a simple JS fiddle does not.

I am still willing to help if you want to reopen and pursue it, but I must do so as a very low priority; I have very real deadlines looming. But if I am the only one who ever encounters it, it seems a waste of time.
